### PR TITLE
Implement new interfaces for ElasticSearch Entity indexing

### PR DESCRIPTION
**Pull request by ko5tik**
Fix https://github.com/jhipster/generator-jhipster/issues/17488

This is updated pull request against issue 17488.   Old PR #18239 was too old and could not be merged  anymore
Everything was cherry picked and some improvements made  to adjust to newew interfaces etc. 

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by ko5tik** [comment]
@DanielFran @mshima   - please review

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 1]
```diff
@@ -29,6 +30,11 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageImpl;
   <%_ } _%>
 import org.springframework.data.domain.Pageable;
+import org.springframework.data.elasticsearch.core.query.NativeSearchQueryBuilder;
+import org.elasticsearch.search.sort.SortBuilder;
+import java.util.List;
+import org.springframework.data.domain.PageRequest;
+
```
Comment (DanielFran) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:37]: Remove extra blank line

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 2]
```diff
@@ -73,10 +85,24 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
+        NativeSearchQuery nativeSearchQuery = queryBuilder.build();
         nativeSearchQuery.setPageable(pageable);
-  <%_ }_%>
+
+<%_ } else { _%>
+        NativeSearchQuery nativeSearchQuery = new NativeSearchQuery(queryStringQuery(query));
+<%_ } _%>
+        return search(nativeSearchQuery);
+    }
+    public Flux<<%= persistClass %>> search(Query query) {
```
Comment (DanielFran) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:104]:
```suggestion

    @Override
    public Flux<<%= persistClass %>> search(Query query) {
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 3]
```diff
@@ -87,39 +113,60 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
 interface <%= entityClass %>SearchRepositoryInternal {
   <%_ if (paginationNo) { _%>
     Stream<<%= persistClass %>> search(String query);
+    Stream<<%= persistClass %>> search(Query query);
```
Comment (DanielFran) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:116]:
```suggestion

    Stream<<%= persistClass %>> search(Query query);
```

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 4]
```diff
@@ -87,39 +113,60 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
   <%_ } else { _%>
     Page<<%= persistClass %>> search(String query, Pageable pageable);
+    Page<<%= persistClass %>> search(Query query);
```
Comment (DanielFran) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:119]:
```suggestion

    Page<<%= persistClass %>> search(Query query);
```

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs [hunk 5]
```diff
@@ -640,8 +643,10 @@ _%>
         List<<%= persistClass %>> <%= entityInstance %>List = <%= entityInstance %>Repository.findAll()<%= callListBlock %>;
         assertThat(<%= entityInstance %>List).hasSize(databaseSizeBeforeCreate + 1);
   <%_ if (searchEngineElasticsearch) { _%>
+      await().atMost(5, TimeUnit.SECONDS).untilAsserted(() ->{
```
Comment (DanielFran) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:646]: There is some issues with indentation, can you run "prettier:format"?

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs [hunk 6]
```diff
@@ -1413,6 +1418,8 @@ _%>
     <%_ } _%>
   <%_ } _%>
   <%_ if (searchEngineElasticsearch) { _%>
+      await().atMost(5, TimeUnit.SECONDS).untilAsserted(() ->{
```
Comment (DanielFran) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:1421]: There is some issues with indentation, can you run "prettier:format"?

File: package.json [hunk 7]
```diff
@@ -126,8 +126,8 @@
     "eslint-config-airbnb-base": "15.0.0",
     "eslint-config-prettier": "8.5.0",
     "eslint-plugin-chai-friendly": "0.7.2",
-    "eslint-plugin-import": "2.26.0",
-    "eslint-plugin-mocha": "10.0.4",
+    "eslint-plugin-import": "2.25.4",
+    "eslint-plugin-mocha": "10.0.3",
```
Comment (DanielFran) [inline package.json:130]: Revert changes

File: package-lock.json [hunk 8]
```diff
@@ -55,8 +55,8 @@
         "eslint-config-airbnb-base": "15.0.0",
         "eslint-config-prettier": "8.5.0",
         "eslint-plugin-chai-friendly": "0.7.2",
-        "eslint-plugin-import": "2.26.0",
-        "eslint-plugin-mocha": "10.0.4",
+        "eslint-plugin-import": "2.25.4",
+        "eslint-plugin-mocha": "10.0.3",
```
Comment (DanielFran) [inline package-lock.json:59]: Revert changes in package-lock.json file

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs [hunk 9]
```diff
@@ -102,6 +102,9 @@ import <%= entityAbsolutePackage %>.service.criteria.<%= entityClass %>Criteria;
 <%_ if (searchEngineElasticsearch) { _%>
 import org.assertj.core.util.IterableUtil;
 import org.apache.commons.collections4.IterableUtils;
+import static org.awaitility.Awaitility.await;
```
Comment (DanielFran) [inline generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs:105]: Static import should be moved at the end of the import list

File: generators/server/templates/build.gradle.ejs [hunk 10]
```diff
@@ -426,6 +426,7 @@ if (SPRING_BOOT_VERSION.indexOf('M') > -1 || SPRING_BOOT_VERSION.indexOf('RC') >
 <%_ } _%>
 <%_ if (searchEngineElasticsearch) { _%>
     testImplementation "org.testcontainers:elasticsearch"
+    testImplementation 'org.awaitility:awaitility:4.2.0'
```
Comment (DanielFran) [inline generators/server/templates/build.gradle.ejs:429]: Add the dependency in pom.xml too

**Review by DanielFran — CHANGES_REQUESTED** [review]
@ko5tik Is it missing the separated transaction?
https://github.com/ko5tik/generator-jhipster/blob/034e69b76dfd7fd64a96b84c112e514f841b0892/generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs#L166-L171

File: generators/server/templates/build.gradle.ejs (hunk 10 above)
Comment (ko5tik) [reply generators/server/templates/build.gradle.ejs:429]: done

File: generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs (hunk 9 above)
Comment (ko5tik) [reply generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs]: done

**Comment by ko5tik** [comment]
> @ko5tik Is it missing the separated transaction? https://github.com/ko5tik/generator-jhipster/blob/034e69b76dfd7fd64a96b84c112e514f841b0892/generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs#L166-L171

Seemd to be not necessary anymore -  I verified it with generated ptoject and itr eindexes necessary dependencies

File: package.json (hunk 7 above)
Comment (ko5tik) [reply package.json]: Reverted,  but  this prevents mocha from running:

ko5tik@roadrunner:~/sources/generator-jhipster$ npm run test

> generator-jhipster@7.8.1 pretest
> npm run lint


> generator-jhipster@7.8.1 lint
> npm run eslint && npm run ejs-lint


> generator-jhipster@7.8.1 eslint
> eslint . --ext .js,.cjs,.mjs


Oops! Something went wrong! :(

ESLint: 8.16.0

Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Failed to load plugin 'mocha' declared in '.eslintrc.json': Package subpath './src/find' is not defined by "exports" in /home/ko5tik/sources/generator-jhipster/node_modules/ramda/package.json
    at new NodeError (node:internal/errors:377:5)
    at throwExportsNotFound (node:internal/modules/esm/resolve:440:9)
    at packageExportsResolve (node:internal/modules/esm/resolve:719:3)
    at resolveExports (node:internal/modules/cjs/loader:483:36)
    at Module._findPath (node:internal/modules/cjs/loader:523:31)
    at Module._resolveFilename (node:internal/modules/cjs/loader:925:27)
    at Module._load (node:internal/modules/cjs/loader:780:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (/home/ko5tik/sources/generator-jhipster/node_modules/v8-compile-cache/v8-compile-cache.js:159:20)
    at Object.<anonymous> (/home/ko5tik/sources/generator-jhipster/node_modules/eslint-plugin-mocha/lib/rules/handle-done-callback.js:3:14)

File: package-lock.json (hunk 8 above)
Comment (ko5tik) [reply package-lock.json]: done

**Comment by DanielFran** [comment]
@ko5tik I added few more changes

**Comment by ko5tik** [comment]
> @ko5tik I added few more changes
> 
> Can you fix those errors: https://github.com/jhipster/generator-jhipster/runs/6610855022?check_suite_focus=true#step:21:77

One fixed,   other  problematic -  SortToSortBuilderListConverter  -   can not find where this is generated and whether this is necessary

**Comment by ko5tik** [comment]
> > @ko5tik I added few more changes
> > Can you fix those errors: https://github.com/jhipster/generator-jhipster/runs/6610855022?check_suite_focus=true#step:21:77
> 
> One fixed, other problematic - SortToSortBuilderListConverter - can not find where this is generated and whether this is necessary

ok, found it -  pulled fresh logic form trunk

**Review by DanielFran — APPROVED** [review]

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 11]
```diff
@@ -87,39 +103,62 @@ class <%= entityClass %>SearchRepositoryInternalImpl implements <%= entityClass
...
             .collect(Collectors.toList());
 
-        return new PageImpl<>(hits, pageable, hits.size());
+        return new PageImpl<>(hits, query.getPageable(), hits.size());
     }
   <%_ } _%>
+
+    @Override
+    public void index(<%= persistClass %> entity) {
+        repository.find<% if (implementsEagerLoadApis ) { %>OneWithEagerRelationships<% } else { %>ById<% } %>(entity.getId()).ifPresent(elasticsearchTemplate::save);
```
Comment (DanielFran) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:160]:
```suggestion
        repository.find<% if (implementsEagerLoadApis) { %>OneWithEagerRelationships<% } else { %>ById<% } %>(entity.getId()).ifPresent(elasticsearchTemplate::save);
```

**Review by DanielFran — CHANGES_REQUESTED** [review]

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs (hunk 11 above)
Comment (ko5tik) [reply generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs]: Seems to be outdated - can not see the difference

Comment (DanielFran) [reply generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs]: Already solved, I committed my suggestion ;)

**Comment by mshima** [comment]
Why the `@Async` was removed?

**Comment by JeroenAdam** [comment]
#16136 #18729

**Comment by RDsideNow** [comment]
I can confirm that as of version 7.8.1 this issue is still present

Screen Shot 2022-05-26 at 6 39 41 PM

If is use the swagger page search, the results return as expected

Screen Shot 2022-05-26 at 6 42 41 PM

Screen Shot 2022-05-26 at 6 42 27 PM

Screen Shot 2022-05-26 at 6 44 39 PM

Screen Shot 2022-05-26 at 6 44 51 PM

**Comment by ko5tik** [comment]
> Why the `@Async` was removed?

I checked   with generated app and @Async   was not necessary for function   and it was causing  problems in tests

**Review by DanielFran — APPROVED** [review]

**Comment by OmarHawk** [comment]
> > #16136 #18729
> 
> This is not going to fix circular indexing issue unfortunately. To break indexing loops you have to place @org.springframework.data.annotation.Transient on the relationships manually - elasticsearch used to honod @JsonIgnoreProperties, which was generated earlier but not anymore.
> 
> And unfortunately you can not add anotations to JDL relationships - only to properties so you have to add it manually.

Wouldn't it make sense to then automatically generate the @Transient attribute on top of the relationship fields so that you wouldn't need to manually place it? Because if you place it on the wrong fields I noticed it will get rid of the some data in db ;-)

**Comment by ko5tik** [comment]
> Wouldn't it make sense to then automatically generate the `@Transient` attribute on top of the relationship fields so that you wouldn't need to manually place it? Because if you place it on the wrong fields I noticed it will get rid of the some data in db ;

It depends.  If indexing does not follow relationship you will be not able to query for those data  - say in your use case  "find me a car which delivered fluid assistance on 25th april 1985"  from a car search.   I think this may be a good default though,  because it will prevent  circular dependencies which are bound to occur pretty fast unless you design your relationship directionality  and navigability  very carefully  - and new users are not doing this  and will stumble upon this issue pretty fast,

**Comment by DanielFran** [comment]
@OmarHawk @ko5tik Can this other issue been take care in another PR?

**Comment by ko5tik** [comment]
> @OmarHawk @ko5tik Can this other issue been take care in another PR?

It should definitely

**Comment by DanielFran** [comment]
Thanks @ko5tik for the contribution

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 12]
```diff
@@ -37,9 +42,13 @@ import org.springframework.data.elasticsearch.core.ReactiveElasticsearchTemplate
...
 <%_ } _%>
 import org.springframework.data.elasticsearch.core.query.NativeSearchQuery;
+import org.springframework.data.elasticsearch.core.query.Query;
 import org.springframework.data.elasticsearch.core.SearchHit;
 import org.springframework.data.elasticsearch.repository.<% if (reactive) {%>Reactive<% } %>ElasticsearchRepository;
-
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.transaction.annotation.Isolation;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:51]: Not used.

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs [hunk 13]
```diff
@@ -29,6 +30,10 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageImpl;
   <%_ } _%>
 import org.springframework.data.domain.Pageable;
+import org.springframework.data.elasticsearch.core.query.NativeSearchQueryBuilder;
+import org.elasticsearch.search.sort.SortBuilder;
+import java.util.List;
+import org.springframework.data.domain.PageRequest;
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:36]: Not used.

**Review by mshima — COMMENTED** [review]
There are imports that aren't used.

**Comment by mshima** [comment]
Sorry for late review.

File: generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs (hunk 13 above)
Comment (ko5tik) [reply generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:36]: This is already merged?   Still possible to use branch?

Comment (DanielFran) [reply generators/entity-server/templates/src/main/java/package/repository/search/EntitySearchRepository.java.ejs:36]: @ko5tik Yes you can

**Pull request merged**